File: project3/xkcdpwgen.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class xkcdpwgen {

  private static List<String> wordList = new ArrayList<>();

  public static void main(String[] args) {
    loadWordList();

    int numWords = 4;
    int numCaps = 0;
    int numNumbers = 0;
    int numSymbols = 0;

    for (int i = 0; i < args.length; i++) {
      if (args[i].equals("-w") | args[i].equals("--words")) {
        numWords = Integer.parseInt(args[i + 1]);
      } else if (args[i].equals("-c") | args[i].equals("--caps")) {
        numCaps = Integer.parseInt(args[i + 1]);
      } else if (args[i].equals("-n") | args[i].equals("--numbers")) {
        numNumbers = Integer.parseInt(args[i + 1]);
      } else if (args[i].equals("-s") | args[i].equals("--symbols")) {
        numSymbols = Integer.parseInt(args[i + 1]);
      } else if (args[i].equals("-h") | args[i].equals("--help")) {
        System.out.println("usage: xkcdpwgen [-h] [-w WORDS] [-c CAPS] [-n NUMBERS] [-s SYMBOLS]\n");
        System.out.println("Generate a secure, memorable password using the XKCD method\n");
        System.out.println("optional arguments:\n");
        System.out.println("    -h, --help            show this help message and exit");
        System.out.println("    -w WORDS, --words WORDS");
        System.out.println("                          include WORDS words in the password (default=4)");
        System.out.println("-c CAPS, --caps CAPS  capitalize the first letter of CAPS random words");
        System.out.println("                          (default=0");
        System.out.println("    -n NUMBERS, --numbers NUMBERS");
        System.out.println("                          insert NUMBERS random numbers in the password");
        System.out.println("                          (default=0)");
        System.out.println("    -s SYMBOLS, --symbols SYMBOLS");
        System.out.println("                          insert SYMBOLS random symbols in the password");
        System.out.println("                          (default=0)");
      }
    }

    generateAndPrintPassword(numWords, numCaps, numNumbers, numSymbols);
  }

  private static void loadWordList() {
    System.out.println(System.getProperty("user.dir"));
    try (BufferedReader reader = new BufferedReader(new FileReader("words.txt"))) {
      String line;
      while ((line = reader.readLine()) != null) {
        wordList.add(line.trim());
      }
    } catch (IOException e) {
      e.printStackTrace();
      System.exit(1);
    }
  }

  private static String generatePassword(int numWords, int numCaps, int numNumbers, int numSymbols) {
    Random random = new Random();
    StringBuilder password = new StringBuilder();

    for (int i = 0; i < numWords; i++) {
      password.append(wordList.get(random.nextInt(wordList.size())));
    }

    for (int i = 0; i < numCaps; i++) {
      int wordIndex = random.nextInt(numWords);
      password.setCharAt(wordIndex, Character.toUpperCase(password.charAt(wordIndex)));
    }

    String symbols = "~!@#$%^&*.:;";
    for (int i = 0; i < numSymbols; i++) {
      int position = random.nextInt(password.length() + 1);
      password.insert(position, symbols.charAt(random.nextInt(symbols.length())));
    }

    for (int i = 0; i < numNumbers; i++) {
      int position = random.nextInt(password.length() + 1);
      password.insert(position, random.nextInt(10));
    }

    return password.toString();
  }

  private static void generateAndPrintPassword(int numWords, int numCaps, int numNumbers, int numSymbols) {
    String password = generatePassword(numWords, numCaps, numNumbers, numSymbols);
    System.out.println(password);
  }
}

File: project3/xkcdpwgen.py
#!/usr/bin/env python3

import argparse
import random

# Load the wordlist from the words.txt file
with open("words.txt", "r") as f:
    wordlist = [line.strip() for line in f]

def generate_password(num_words, num_caps, num_numbers, num_symbols):
    password_words = random.sample(wordlist, num_words)
    password = "".join(password_words)

    # Capitalize random words
    for _ in range(num_caps):
        word_index = random.randint(0, num_words - 1)
        password = password[:word_index] + password[word_index].capitalize() + password[word_index + 1:]

    # Insert random numbers
    for _ in range(num_numbers):
        position = random.randint(0, len(password))
        password = password[:position] + str(random.randint(0, 9)) + password[position:]

    # Insert random symbols
    symbols = "~!@#$%^&*.:;"
    for _ in range(num_symbols):
        position = random.randint(0, len(password))
        symbol = random.choice(symbols)
        password = password[:position] + symbol + password[position:]

    return password

def main():
    parser = argparse.ArgumentParser(description="Generate a secure, memorable password using the XKCD method")
    parser.add_argument("-w", "--words", type=int, default=4, help="include WORDS words in the password (default=4)")
    parser.add_argument("-c", "--caps", type=int, default=0, help="capitalize the first letter of CAPS random words (default=0)")
    parser.add_argument("-n", "--numbers", type=int, default=0, help="insert NUMBERS random numbers in the password (default=0)")
    parser.add_argument("-s", "--symbols", type=int, default=0, help="insert SYMBOLS random symbols in the password (default=0)")
    args = parser.parse_args()

    if args.words <= 0:
        print("Number of words must be greater than 0.")
        return

    if args.caps < 0:
        print("Number of capitalized words cannot be negative.")
        return

    if args.numbers < 0:
        print("Number of numbers cannot be negative.")
        return

    if args.symbols < 0:
        print("Number of symbols cannot be negative.")
        return

    if args.words > len(wordlist):
        print("Not enough words in the wordlist.")
        return

    password = generate_password(args.words, args.caps, args.numbers, args.symbols)
    print(password)

if __name__ == "__main__":
    main()

